zulip_bots/zulip_bots/bots/connect_four/connect_four.py

```diff
@@ -0,0 +1,535 @@
+
+    # Updates to the main bot handler that all state managers must use
+    def basic_updates(self):
+        if self.users is not None:
+            self.main_bot_handler.inputVerification.verified_users = self.users
+
+        if self.state:
+            self.main_bot_handler.state = self.state
+
+        self.main_bot_handler.user_messages = self.user_messages
+
+        self.main_bot_handler.opponent_messages = self.opponent_messages
+
+    def reset_self(self):
+        self.users = None
+        self.user_messages = []
+        self.opponent_messages = []
+        self.state = ''
+
+class GameCreator(StateManager):
+    def __init__(self, main_bot_handler):
+        super(GameCreator, self).__init__(main_bot_handler)
+        self.gameHandler = None
+        self.invitationHandler = None
+
+    def handle_message(self, content, sender):
+        if content == 'start game with computer':
+            self.users = [sender]
+            self.state = 'playing'
+            self.gameHandler = GameHandler(self.main_bot_handler, 'one_player')
+
+            self.user_messages.append('**You started a new game with the computer!**')
+            self.user_messages.append(self.gameHandler.parse_board())
+            self.user_messages.append(self.gameHandler.your_turn_message())
+
+        elif re.compile('\w+@\w+\.\w+').search(content):
+            opponent = re.compile('(\w+@\w+\.\w+)').search(content).group(1)
+
+            if opponent == sender:
+                self.user_messages.append('You can\'t play against yourself!')
+                self.update_main_bot_handler()
+                return
+
+            self.users = [sender, opponent]
+            self.state = 'inviting'
+            self.gameHandler = GameHandler(self.main_bot_handler, 'two_player')
+            self.invitationHandler = InvitationHandler(self.main_bot_handler)
+
+            self.user_messages.append(self.invitationHandler.confirm_new_invitation(opponent))
+
+            self.opponent_messages.append(self.invitationHandler.alert_new_invitation(sender))
+
+        self.update_main_bot_handler()
+
+    def update_main_bot_handler(self):
+        self.basic_updates()
+
+        self.main_bot_handler.player_cache = self.users
+
+        self.main_bot_handler.gameHandler = deepcopy(self.gameHandler)
+
+        if self.invitationHandler:
+            self.main_bot_handler.invitationHandler = deepcopy(self.invitationHandler)
+
+        self.reset_self()
+
+class GameHandler(StateManager):
+    def __init__(self, main_bot_handler, game_type, board = ConnectFourModel().blank_board, turn = 0):
+        super(GameHandler, self).__init__(main_bot_handler)
+        self.game_type = game_type
+        self.board = board
+        self.turn = turn
+        self.game_ended = False
+        self.connectFourModel = ConnectFourModel()
+        self.connectFourModel.update_board(board)
+        self.tokens = [':blue_circle:', ':red_circle:']
+
+    def parse_board(self):
```

Comment:
So if you really wanted to take these concepts further, I would argue that you could move `parse_board` into a ConnectFour specific class, and leave this class to really just handle the concept of a two-player game generically.